### PR TITLE
Send emails when auction ends

### PR DESCRIPTION
**Pull request by adelevie**
History: https://github.com/18F/micropurchase/pull/797

Accidentally merged that PR ^ then reverted and brought it back to life here.

### CODE REVIEW

File: features/step_definitions/auction_create_steps.rb [hunk 1]
```diff
@@ -10,6 +10,22 @@
   @auction = FactoryGirl.create(:auction, :closed, :with_bidders)
 end
 
+Given(/^I have won an auction$/) do
+  @auction = FactoryGirl.build(:auction, :closed, :with_bidders)
+  bids = @auction.bids.sort_by(&:amount)
+  b = bids.first
```
Comment (jessieay) [inline features/step_definitions/auction_create_steps.rb]: can we use `WinningBid` here or do something similar so we are *sure* this is the winning bid?

File: features/step_definitions/auction_create_steps.rb [hunk 2]
```diff
@@ -10,6 +10,22 @@
...
+  b.update_attribute(:bidder_id, @user.id)
+  SaveAuction.new(@auction).perform
+end
+
+Given(/^I have lost an auction$/) do
+  @auction = FactoryGirl.build(:auction, :closed, :with_bidders)
+  bids = @auction.bids.sort_by(&:amount)
+  b = bids.last
```
Comment (jessieay) [inline features/step_definitions/auction_create_steps.rb]: same as above: let's be sure this is the losing bid

File: app/services/save_auction.rb [hunk 3]
```diff
@@ -4,10 +4,20 @@ def initialize(auction)
...
   end
 
   private
 
   attr_reader :auction
+
+  def schedule_auction_ended_job(saved)
+    if saved
+      job = AuctionEndedJob.new(auction.id)
+      Delayed::Job.enqueue(job, run_at: auction.ended_at)
```
Comment (jessieay) [inline app/services/save_auction.rb]: do we want to add logic for updating this job if `ended_at` changes or is that for a separate PR?

File: app/services/save_auction.rb [hunk 4]
```diff
@@ -4,10 +4,20 @@ def initialize(auction)
   end
 
   def perform
-    auction.save
+    saved = auction.save
+    schedule_auction_ended_job(saved)
```
Comment (jessieay) [inline app/services/save_auction.rb]: thoughts on passing `auction` and checking if it is persisted down on L18? I know `save` returns true or false but truthiness and falsiness sometimes get me into weird places. Clearer to check for actual values / methods like `persisted?`

File: features/vendor_views_closed_auction.feature [hunk 5]
```diff
@@ -4,29 +4,25 @@ Feature: Vendor views closed auctions
...
 
   Scenario: I am the winner
-    Given there is a closed auction
-    And I am an authenticated vendor
-    And I have placed the lowest bid
+    Given I am an authenticated vendor
+    And I have won an auction
     When I visit the auction page
     Then I should see the auction had a winning bid with name
     And I should see I am the winner
```
Comment (jessieay) [inline features/vendor_views_closed_auction.feature]: thoughts on how we can enhance this test for new logic:

```
Scenario: I am the winner
    Given I am an authenticated vendor
    And I have won an auction
    And the auction has not ended
    Then I should not have received an email notifying me that I won
    When the auction ends # this step includes time travel
    And I visit the auction page
    Then I should see the auction had a winning bid with name
    And I should see I am the winner
    And I should see when the auction started
    And I should see when the auction ended
    And there should be meta tags for the closed auction
    And I should receive an email notifying me that I won
```

**Comment by adelevie** [comment]
all shiny and ready

File: app/jobs/delayed_job.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class Delayed::Job < ActiveRecord::Base
+  self.table_name = 'delayed_jobs'
+  attr_accessible :auction_id
```
Comment (jessieay) [inline app/jobs/delayed_job.rb]: do we need this? I think we get this for free from database

File: app/jobs/delayed_job.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
+  belongs_to :auction, polymorphic: true
```
Comment (jessieay) [inline app/jobs/delayed_job.rb]: do we need this?

File: app/services/save_auction.rb [hunk 8]
```diff
@@ -5,9 +5,22 @@ def initialize(auction)
...
   private
 
   attr_reader :auction
+
+  def schedule_auction_ended_job
+    if auction.persisted?
+      job = AuctionEndedJob.new(auction.id)
+      Delayed::Job.enqueue(job,
+                           run_at: auction.ended_at,
+                           queue: 'auction_ended',
```
Comment (jessieay) [inline app/services/save_auction.rb]: 👏

File: app/services/update_auction.rb [hunk 9]
```diff
@@ -21,6 +23,31 @@ def perform
...
+    if updating_ended_at?
+      job = auction_ended_job
+
+      if job
+        job.update(run_at: auction.ended_at)
+      end
+    end
+  end
+
+  def auction_ended_job
```
Comment (jessieay) [inline app/services/update_auction.rb]: so fancy 💯

File: app/services/winning_bidder_email_sender.rb [hunk 10]
```diff
@@ -4,15 +4,21 @@ def initialize(auction)
...
+        .deliver_later
+    end
   end
 
   private
 
   attr_reader :auction
 
+  def auction_has_winner?
+    !WinningBid.new(auction).find.nil?
```
Comment (jessieay) [inline app/services/winning_bidder_email_sender.rb]: `!WinningBid.new(auction).find.nil?` == `WinningBid.new(auction).find.present?` (latter may be easier to understand...)

File: spec/jobs/auction_ended_job_spec.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+        expect(winning_bidder_email_sender_double).to receive(:perform)
+
+        AuctionEndedJob.new(auction.id).perform
+      end
+
+      it 'sends the losing bidders email' do
+        auction = create(:auction)
+
+        auction = create(:auction, :closed, :with_bidders)
```
Comment (jessieay) [inline spec/jobs/auction_ended_job_spec.rb]: we've got two `auction` vars in this test - do we need both?

File: spec/services/update_auction_spec.rb [hunk 12]
```diff
@@ -2,6 +2,35 @@
...
+        parsed_new_ended_at = DateTimeParser.new(new_ended_at, 'ended_at').parse
+        params = { auction: new_ended_at}
+
+        expect(job.run_at).to eq(auction.ended_at)
+        UpdateAuction.new(auction: auction,
+                          params: params,
+                          current_user: auction.user).perform
+
+        job.reload
+        expect(job.run_at).to eq(parsed_new_ended_at)
```
Comment (jessieay) [inline spec/services/update_auction_spec.rb]: WONDERFUL

**Comment by jessieay** [comment]
final few 👶  comments -- looking 🌈 🌈

**Comment by adelevie** [comment]
I'll get those tomorrow. Can't wait to try this on staging.

**Comment by jessieay** [comment]
LOOKS AMAZING 💭

**Pull request merged**